napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,212 @@
+"""Plugin related functions that require Qt.
+
+Non-Qt plugin functions can be found in: `napari/plugins/_npe2.py`
+"""
+from __future__ import annotations
+
+import inspect
+from typing import (
+    TYPE_CHECKING,
+    List,
+    NamedTuple,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+
+from app_model import Action
+from app_model.types import SubmenuItem, ToggleRule
+from magicgui.type_map._magicgui import MagicFactory
+from magicgui.widgets import FunctionGui, Widget
+from npe2 import plugin_manager as pm
+from qtpy.QtWidgets import QWidget
+
+from napari._app_model import get_app
+from napari._app_model.constants import MenuGroup, MenuId
+from napari._qt._qapp_model.qactions._qproviders import _provide_window
+from napari.plugins import menu_item_template
+from napari.plugins._npe2 import (
+    _get_contrib_parent_menu,
+    get_widget_contribution,
+)
+from napari.utils.translations import trans
+from napari.viewer import Viewer
+
+if TYPE_CHECKING:
+    from npe2.manifest import PluginManifest
+    from npe2.plugin_manager import PluginName
+    from npe2.types import WidgetCreator
+
+    from napari._qt.qt_main_window import Window
+
+
+def _get_widget_viewer_param(
+    widget_callable: WidgetCreator, widget_name: str
+) -> str:
+    """Get widget parameter name (if any) and check type."""
+    if inspect.isclass(widget_callable) and issubclass(
+        widget_callable,
+        (QWidget, Widget),
+    ):
+        widget_param = ""
+        # Inspection can fail when adding to bundle as it thinks widget is
+        # a builtin
+        try:
+            sig = inspect.signature(widget_callable.__init__)
+        except ValueError:
+            pass
+        else:
+            for param in sig.parameters.values():
+                if param.name == 'napari_viewer' or param.annotation in (
+                    'napari.viewer.Viewer',
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    elif isinstance(widget_callable, MagicFactory) or inspect.isfunction(
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "'{widget}' must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or function. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+            )
+        )
+    return widget_param
+
+
+def _get_widget_callback(
```

Comment:
nit: is this actually instantiating the widget? Because if so, then it's not getting a callback right?

Reply:
Sorry I don't follow, a callback can't instantiate the widget? I was just meaning it's to be used in the `callback` field...?

Reply:
Also it doesn't always instantiate, it toggles if the widget already exists..

Reply:
Ah yes @lucyleeow I think I got confused. Ok super nitty, but I think it should be called `_get_widget` in that case.